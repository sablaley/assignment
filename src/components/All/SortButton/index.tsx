import React from "react";
import SortButton from "./SortButton";
export type SortType = "name" | "population" | "area";

interface Props {
  readonly type: SortType;
  readonly currentSort?: SortType;
  readonly sortAsc: boolean;
  readonly setCurrentSort: (newSort: SortType) => void;
}

export default ({type, currentSort, sortAsc, setCurrentSort}: Props): JSX.Element => {
  const [asc, setAsc] = React.useState<boolean>();

  const active = currentSort === type;

  React.useEffect(() => {
    active && setAsc(sortAsc);
  }, [sortAsc, active]);

  return <SortButton active={active} asc={asc} onClick={() => setCurrentSort(type)}></SortButton>;
};
