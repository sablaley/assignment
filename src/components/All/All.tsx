import * as React from "react";
import {Country,Population,Area} from "../../models/Country";
import SortButton, {SortType} from "./SortButton";

export interface IAllProps {
  readonly countries: Array<Country>;
  readonly population: Array<Population>;
  readonly area: Array<Area>;
  readonly currentSort?: SortType;
  readonly setCurrentSort: (_: SortType) => void;
  readonly sortAsc: boolean;
  
}

export const All = (props: IAllProps): JSX.Element => (
   <div className="all-cities-table">
     <table>
       <thead>
         <tr>
           <th>
             Name <SortButton type="name" {...props} ></SortButton>
           </th>
           <th>
             Population <SortButton type="population" {...props} ></SortButton>
           </th>
           <th>
             Area <SortButton type="area" {...props}></SortButton>
           </th>
         </tr>
       </thead>
       <tbody>
          {props.population.map(item1 => {
           const matchingItem = props.area.find(item2 => item2.country === item1.country);
           return (
             <tr key={item1.country}>
               <td> {item1.country}</td>
               <td>{item1.population}</td>
               <td>{matchingItem ? matchingItem.area : 'N/A'}</td>
             </tr>
           );
         })}
      </tbody>
     </table>
   </div>
 );
