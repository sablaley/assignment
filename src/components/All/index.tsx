import {All} from "@components/All/All";
import React from "react";
import {SortType} from "./SortButton";
const countryByName = require("../../../country-json/src/country-by-name.json");
const populationByCountry = require("../../../country-json/src/country-by-population.json");
const populationByArea = require("../../../country-json/src/country-by-surface-area.json");

export default (): JSX.Element => {
  const [currentSort, setCurrentSort] = React.useState<SortType>();
  const [sortAsc, toggleSortAsc] = React.useState<boolean>(true);

  const handleSortUpdate = (newType: SortType) => {
    if (newType === currentSort) {
      toggleSortAsc(!sortAsc);
    }
    setCurrentSort(newType);
  };

  return (
    <All currentSort={currentSort} sortAsc={sortAsc} countries={countryByName} setCurrentSort={handleSortUpdate}  population={populationByCountry} area={populationByArea}/>
  );
};
