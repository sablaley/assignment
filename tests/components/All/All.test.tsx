import {All, IAllProps} from "@components/All/All";
import React from "react";
import renderer from "react-test-renderer";

describe("All component", () => {
  test("should render component properly", () => {
    // given
    const dummyComponentProps: IAllProps = {
      currentSort: undefined,
      sortAsc: false,
      countries: [],
      population:[],
      area:[],
      setCurrentSort: jest.fn(),
    };

    // when
    const componentRenderer = renderer.create(<All {...dummyComponentProps} />);
    const tree = componentRenderer.toJSON();

    // then
    expect(tree).toMatchSnapshot();
  });
});
