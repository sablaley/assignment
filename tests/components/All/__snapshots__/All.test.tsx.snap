// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All component should render component properly 1`] = `
<div
  className="all-cities-table"
>
  <table>
    <thead>
      <tr>
        <th>
          Name 
          <a
            className="sort-button"
            onClick={[Function]}
          >
            ▼
          </a>
        </th>
        <th>
          Population 
          <a
            className="sort-button"
            onClick={[Function]}
          >
            ▼
          </a>
        </th>
        <th>
          Area 
          <a
            className="sort-button"
            onClick={[Function]}
          >
            ▼
          </a>
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
